tools/prepare-fork-repo/main.go

```diff
@@ -0,0 +1,116 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"log"
+	"time"
+
+	"gopkg.in/src-d/go-billy.v3/memfs"
+	git "gopkg.in/src-d/go-git.v4"
+	gitconfig "gopkg.in/src-d/go-git.v4/config"
+	"gopkg.in/src-d/go-git.v4/plumbing"
+	"gopkg.in/src-d/go-git.v4/plumbing/object"
+	"gopkg.in/src-d/go-git.v4/plumbing/storer"
+	"gopkg.in/src-d/go-git.v4/storage/memory"
+)
+
+func remoteBranches(s storer.ReferenceStorer) (storer.ReferenceIter, error) {
+	refs, err := s.IterReferences()
+	if err != nil {
+		return nil, err
+	}
+
+	return storer.NewReferenceFilteredIter(func(ref *plumbing.Reference) bool {
+		return ref.Name().IsRemote()
+	}, refs), nil
+}
+
+func main() {
+	repoName := flag.String("repo", "", "origin fork repository name (eg. 'containers-image')")
+	baseBranchName := flag.String("base-branch", "", "base branch name (eg. 'openshift-3.9')")
+	branchName := flag.String("branch", "", "new branch name (eg. 'openshift-3.10')")
+	initialCommit := flag.String("commit", "", "a starting point commit hash")
+	flag.Parse()
+	if len(*repoName) == 0 {
+		log.Fatalf("flag -repo must be set")
+	}
+	if len(*branchName) == 0 {
+		log.Fatalf("flag -branch must be set")
+	}
+	if len(*initialCommit) == 0 {
+		log.Fatalf("flag -commit must be set")
+	}
+	if len(*baseBranchName) == 0 {
+		log.Fatalf("flag -base-branch must be set")
+	}
+
+	log.Printf("--> cloning openshift/%s ...", *repoName)
+	r, err := git.Clone(memory.NewStorage(), memfs.New(), &git.CloneOptions{
+		URL: "git@github.com:openshift/" + *repoName,
+	})
+	if err != nil {
+		log.Fatalf("error cloning %s: %v", repoName, err)
+	}
+	bs, err := remoteBranches(r.Storer)
+	if err != nil {
+		log.Fatalf("unable to list remote branches: %v", err)
+	}
+	var (
+		baseBranchHash plumbing.Hash
+		dstBranch      *plumbing.Reference
+	)
+	err = bs.ForEach(func(ref *plumbing.Reference) error {
+		if ref.Name().Short() == "origin/"+*branchName {
+			dstBranch = ref
+		}
+		if ref.Name().Short() == "origin/"+*baseBranchName {
+			baseBranchHash = ref.Hash()
+		}
+		return nil
+	})
+	if dstBranch != nil {
+		log.Printf("--> branch %s already exists", *branchName)
+	} else {
+		log.Printf("--> creating new branch %s", *branchName)
+		dstBranch = plumbing.NewHashReference(plumbing.ReferenceName(fmt.Sprintf("refs/heads/origin/%s", *branchName)), baseBranchHash)
+	}
+	if err := r.Storer.SetReference(dstBranch); err != nil {
+		log.Fatalf("unable to create %s branch: %v", *branchName, err)
+	}
+	if baseBranchHash.IsZero() {
+		log.Fatalf("base branch %s does not exists", baseBranchName)
+	}
+
+	w, _ := r.Worktree()
+
+	commit, err := w.Commit(fmt.Sprintf("Origin-commit: %s", *initialCommit), &git.CommitOptions{
```

Comment:
you don't even need worktree for this. Just create a commit object directly and store it.

Reply:
I tried but src-d/git complained that you cannot create commit on in-memory storage.